<a href="https://colab.research.google.com/github/Dikk674567456/PROJECT-2.-SQL-/blob/main/PROJECT_2_SQL_%D0%90%D0%BD%D0%B0%D0%BB%D0%B8%D0%B7_%D0%B2%D0%B0%D0%BA%D0%B0%D0%BD%D1%81%D0%B8%D0%B9_%D0%B8%D0%B7_HeadHunter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
# !pip install psycopg2
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python

In [3]:
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432

)

'''
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432
'''

"\nDBNAME = 'project_sql'\nUSER = 'skillfactory'\nPASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'\nHOST = '84.201.134.129'\nPORT = 5432\n"

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [4]:
# текст запроса
query_3_1 = f'''
select count(*) from vacancies
'''

In [5]:
# результат запроса

df = pd.read_sql_query(query_3_1, connection)
display(df)


<ipython-input-5-334c39270908>:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [6]:
# текст запроса
query_3_2 = f'''
select count(*) from employers
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
display(df)


<ipython-input-7-b961e7f54253>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select count(*) from areas
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
display(df)

<ipython-input-9-5723407fa77f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select count(*) from industries
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
display(df)

<ipython-input-11-059deb18e3a2>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# **Выводы по предварительному анализу данных**

3.1 Количество вакансий в базе - 49197<br>
3.2 Работодателей - 23501<br>
3.3 Регионов  - 1362<br>
3.4 Сфер деятельности - 294

In [12]:
# выводы по предварительному анализу данных

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_4_1 = f'''
select
b.name,
count(b.id)
from vacancies as a
left join areas as b on  b.id=a.area_id
group by b.name
order by count(b.id) desc
limit 5
'''



In [14]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
display(df)

<ipython-input-14-75862acd271f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = f'''
select
count(*)
from vacancies as a
where a.salary_from is not null or a.salary_to is not null
'''

In [16]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
display(df)

<ipython-input-16-468edde68b33>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3 = f'''
select
round(avg(salary_from),0),
round(avg(salary_to),0)
from vacancies
'''


In [18]:

# результат запроса
df = pd.read_sql_query(query_4_3, connection)
display(df)

<ipython-input-18-491d2f469024>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


round     round
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:

# текст запроса
query_4_4 = f'''
select
schedule,
employment,
count(id)
from vacancies
group by schedule, employment
order by count(id) desc
'''

In [20]:
# результат запроса

df = pd.read_sql_query(query_4_4, connection)
display(df)


<ipython-input-20-ce3e44934d1e>:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [21]:

# текст запроса
query_4_5 = f'''
select
experience,
count(id)
from vacancies
group by experience
order by count(id)
'''

In [22]:
# результат запроса

df = pd.read_sql_query(query_4_5, connection)
display(df)

<ipython-input-22-b40f9d5b9a78>:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# **Выводы по детальному анализу вакансий**

4.1 Лидеры по кол-ву вакансий<br> - Москва (5333)<br> - Санкт-Петербург (2851)<br> - Минск (2112)<br> - Новосибирск (2006)<br> - Алматы (1892)<br>
4.2 Зарплаты указаны - в 24073 вакансиях<br>
4.3 Средние значения вилки ЗП - 71065 - 110537 руб.<br>
4.4 Лидеры по кол-ву вакансий  сочетания типа рабочего графика (schedule) и типа трудоустройства:<br>
 - Полный день-Полная занятость	 - 35367<br>
 - Удаленная работа-Полная занятость - 7802<br>
 - Гибкий график-Полная занятость - 1593<br>
 4.5 Кол-во вакансий в разрезе опыта:<br>
 - Более 6 лет - 1337<br>
 - Нет опыта - 7197<br>
 - От 3 до 6 лет - 14511<br>
 - От 1 года до 3 лет - 26152<br>

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
# текст запроса
# текст запроса
query_5_1 = f'''
select
b.name,
count(b.id)
from vacancies
left join employers as b on b.id=vacancies.employer_id
group by b.name
order by count(b.id) desc
limit 5
'''

In [24]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
display(df)

<ipython-input-24-53e9195103b7>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [25]:
# текст запроса
query_5_2 = f'''
SELECT
    a.name AS area_name,
    COUNT(DISTINCT e.id) AS employer_count,
    COUNT(v.id) AS vacancy_count
FROM areas a
LEFT JOIN employers e ON a.id = e.area
LEFT JOIN vacancies v ON a.id = v.area_id
where v.id is null
GROUP BY a.name
ORDER BY employer_count DESC
limit 5
'''

In [26]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
display(df)

<ipython-input-26-69bfba82065b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area_name  employer_count  vacancy_count
0              Россия             410              0
1           Казахстан             207              0
2  Московская область              75              0
3  Краснодарский край              19              0
4            Беларусь              18              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [27]:
# текст запроса
query_5_3 = f'''
SELECT
    a.name,
    COUNT(DISTINCT c.area_id)

FROM employers a
LEFT JOIN vacancies c ON c.employer_id = a.id

GROUP BY a.name
ORDER BY COUNT(DISTINCT c.area_id) desc
limit 5
'''

In [28]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
display(df)

<ipython-input-28-af8d97c75539>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                  Яндекс    181
1              Ростелеком    152
2              Спецремонт    116
3  Поляков Денис Иванович     88
4               ООО ЕФИН      71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [29]:
# текст запроса
query_5_4 = f'''
SELECT

    COUNT(a.id)

FROM employers a
LEFT JOIN employers_industries b ON b.employer_id = a.id

where b.industry_id is null


limit 5
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
display(df)

<ipython-input-30-ba270b911e7e>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [31]:
# текст запроса
query_5_5 = f'''
SELECT
    a.name,
    COUNT(b.employer_id)

FROM employers a
LEFT JOIN employers_industries b ON b.employer_id = a.id
group by a.name
having COUNT(b.employer_id)=4
order by a.name
limit 3
'''

In [32]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
display(df)

<ipython-input-32-dac110f9cb78>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  101 Интернет      4
1      21vek.by      4
2          2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [33]:
# текст запроса
query_5_6 = f'''
SELECT

COUNT(c.id)

FROM employers a
LEFT JOIN employers_industries b ON b.employer_id = a.id
left join industries c on c.id=b.industry_id
where c.name='Разработка программного обеспечения'
'''

In [34]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
display(df)


<ipython-input-34-abda59a868bb>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [35]:
# код для получения списка городов-милионников

import requests
from bs4 import BeautifulSoup

# Загружаем страницу
url = "https://ru.wikipedia.org/wiki/Города-миллионеры_России"
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')

# Ищем таблицу с городами (может потребоваться уточнение селектора)
table = soup.find('table', {'class': 'standard sortable'})

# Парсим данные из таблицы
cities = []
if table:
    for row in table.find_all('tr')[1:]:  # Пропускаем заголовок
        cols = row.find_all('td')
        if len(cols) >= 2:
            # Извлекаем название города
            city_cell = cols[1].find('a')
            city = city_cell.text.strip()
            cities.append(city)

cities=tuple(cities)

display(cities)


('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Краснодар',
 'Самара',
 'Ростов-на-Дону',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [36]:
# текст запроса
query_5_7 = f'''
SELECT

c.name, count(b.id) as vacancy_count

FROM employers a
LEFT JOIN vacancies b ON b.employer_id = a.id
left join areas c on c.id=b.area_id
where a.name='Яндекс' and c.name in {cities}
group by c.name
having count(b.id)<>0
union all
select
'Total',
count(b.id) as vacancy_count

FROM employers a
LEFT JOIN vacancies b ON b.employer_id = a.id
left join areas c on c.id=b.area_id
where a.name='Яндекс' and c.name in {cities}



order by vacancy_count

'''

In [37]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
display(df)

<ipython-input-37-7619c91123bd>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  vacancy_count
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

***

# **Выводы по анализу работодателей**

5.1 Компании лидеры по кол-ву вакансий:<br>
- Яндекс - 1933<br>
- Ростелеком - 491<br>
- Тинькофф - 444<br>
- Сбер - 428<br>
- Газпром-нефть - 331<br>

5.2 Регионы, в которых нет вакансий, с наибольшим кол-вом работодателей:<br>
 - Россия - 410<br>
 - Казахстан - 207<br>
 - МО - 75<br>

5.3 Компании-лидеры по кол-ву регионов с вакансиями<br>
- Яндекс - 181<br>
- Ростелеком - 152<br>
- Спецремонт - 116<br>
- Поляков Денис Иванович - 88<br>
- ООО ЕФИН - 71<br>-

5.4 Кол-во работодателей, у которых не заполненна сфера деятельности - 8419<br>
5.5 Топ компаний, у которых указано 4 вида деятельности:<br>
 - 101 Интернет<br>
 - 21vek.by<br>
 - 2 ГИС<br>
 5.6 Кол-во работодателей, которые занимаются разработкой - 3553<br>
5.7 Кол-во вакансий Яндекса в городах миллионниках - 485 (в лидерах Мск, Питер, ЕК)

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [38]:
# текст запроса
query_6_1 = f'''
SELECT

COUNT(a.id)

FROM vacancies a
where lower(a.name) like '%data%' or lower(a.name) like '%данн%'
'''


In [39]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
display(df)

<ipython-input-39-60087489a893>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [53]:


# текст запроса

query_6_2 = f'''
SELECT

COUNT(a.id)

FROM vacancies a

where (a.name iLIKE '%data scientist%'
  OR a.name iLIKE '%data science%'
  OR a.name iLIKE '%исследователь данных%'
  OR a.name iLIKE '%ML%'
  OR a.name iLIKE '%machine learning%'
  OR a.name iLIKE '%машинн%обучен%')
  and (a.name not ilike '%HTML%')
  and
  (a.name ilike '%junior%'
  or a.experience='Нет опыта'
  or a.employment='Стажировка')
'''

In [54]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
display(df)

<ipython-input-54-3f9976715ced>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [42]:
# текст запроса
query_6_3 = f'''
SELECT

COUNT(a.id)

FROM vacancies a

where (a.name iLIKE '%data scientist%'
  OR a.name iLIKE '%data science%'
  OR a.name iLIKE '%исследователь данных%'
  OR a.name iLIKE '%ML%'
  OR a.name iLIKE '%machine learning%'
  OR a.name iLIKE '%машинн%обучен%')
  and (a.name not ilike '%HTML%')
  and
  (a.key_skills ilike '%sql%'
  or a.key_skills ilike '%postgres%')
'''

In [43]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
display(df)

<ipython-input-43-5a5376965758>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [44]:
# текст запроса
query_6_4 = f'''
SELECT

COUNT(a.id)

FROM vacancies a

where (a.name iLIKE '%data scientist%'
  OR a.name iLIKE '%data science%'
  OR a.name iLIKE '%исследователь данных%'
  OR a.name iLIKE '%ML%'
  OR a.name iLIKE '%machine learning%'
  OR a.name iLIKE '%машинн%обучен%')
  and (a.name not ilike '%HTML%')
  and
  a.key_skills ilike '%python%'

'''

In [45]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
display(df)

<ipython-input-45-f9c1bdabae45>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [46]:
# текст запроса

query_6_5 = f'''
SELECT
round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1),2)

FROM vacancies a
where (a.name iLIKE '%data scientist%'
  OR a.name iLIKE '%data science%'
  OR a.name iLIKE '%исследователь данных%'
  OR a.name LIKE '%ML%'
  OR a.name iLIKE '%machine learning%'
  OR a.name iLIKE '%машинн%обучен%')
  and (a.name not like '%HTML%')
  AND (key_skills IS NOT NULL)




'''

In [47]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
display(df)


<ipython-input-47-48c58f606c99>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [48]:

# текст запроса

query_6_6 = f'''
SELECT

a.experience,
round(avg(coalesce((a.salary_from+a.salary_to)/2,a.salary_from,a.salary_to)))

FROM vacancies a
where
 (a.name iLIKE '%data scientist%'
  OR a.name iLIKE '%data science%'
  OR a.name iLIKE '%исследователь данных%'
  OR a.name LIKE '%ML%'
  OR a.name iLIKE '%machine learning%'
  OR a.name iLIKE '%машинн%обучен%')
  and (a.name not like '%HTML%')

group by a.experience
order by 1

'''



In [49]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
display(df)

<ipython-input-49-f73988b941a8>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

# **Выводы по предметному анализу**


6.1 Вакансий, имеющих отношение к данным - 1771<br>
6.2 Вакансий DS для джуна - 51<br>
6.3 Вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres - 229<br>
6.4 Вакансий для DS с требованием знать Питон - 357<br>
6.5 Ключевых навыков в среднем указывают в вакансиях для DS  - 6.41 шт.<br>
6.6 Средняя ЗП для DS в разрезе опыта:<br>
- 6-... лет - то ли не требуются такие специалисты, так как с таким опытом должны уйти на руководителей, или не указывают ЗП, так как для таких специалистов все сильно индивидуально. В общем тема для доп. исследования<br>
- 3-6 лет - 243115 руб.<br>
- 1-3 года - 146704 руб.<br>
- Нет опыта - 74643 руб.



# **Подведем итог исследования**

1. Лидеры по кол-ву вакансий - крупные города (Москва, Питер, Минск, Новосиб, Алматы)<br>
2. Чаще всего ищут сотрудников с опытом 1-3 года (возможно значительная часть для быстро обучаемых профессий), затем с опытом 3-6 лет<br>
3. Зарплаты указаны только примерно в половине вакансий. У трети компаний не заполнена сфера деятельности<br>
4. Похоже есть пробел в данных и по вакансиям, так так есть крупные регионы, в которых нет вакансий, но есть работодатели (Регионы, в которых нет вакансий (Россия - 410 компаний, Казахстан - 207, МО - 75)<br>
5. Наиболее востребованна  полная занятость, форматы  - Полный день-Полная занятость и Удаленная работа-Полная занятость<br>
6. Компании лидеры по кол-ву вакансий - ИТ (Яндекс), два банка (Т и Сбер) и телеком (Ростелек)<br>
7. При этом по широте регионов, где опубликованы вакансии, первые две срочки в лидерах все те же Яндекс и Ростелек, а вот дальше Спецремонт, некий ИП Поляков Д.И. и ООО ЕФИН<br>
8. Из порядка 24 тыс работодателей с 49 тыс вакансий, разработкой занимается только порядка 3.5 тыс компаний<br>
9. Вакансий, имеющих отношение к данным порядка 1700 шт.<br>
10. Джунов датасаентистов сильно не любят, всего 51 шт. из 536 вакансий по дата саенс<br>
11. Среди требований к саентистам чаще встречается знание Питона (357), чем языки sql запросов (229)<br>
12. Разница в средней ЗП датасаентиста с опытом 1-3 года и 3-6 раз - больше, чем в 1.5 раза (146704 руб. против 243115 руб.)











In [64]:
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

# Сильно любопытно, что за оптимизаторв налогов ИП Поляков, посмотрим вакансии и сферу деятельности

query_7_1 = f'''
SELECT
    c.name,
    COUNT(DISTINCT c.area_id)

FROM employers a
LEFT JOIN vacancies c ON c.employer_id = a.id
where a.name='Поляков Денис Иванович'
GROUP BY c.name
ORDER BY COUNT(DISTINCT c.area_id) desc
limit 5
'''
df = pd.read_sql_query(query_7_1, connection)
display(df)

query_7_2 = f'''
SELECT
    c.name


FROM employers a
LEFT JOIN employers_industries b ON b.employer_id = a.id
LEFT JOIN industries c ON b.industry_id = c.id
where a.name='Поляков Денис Иванович'


'''
df = pd.read_sql_query(query_7_2, connection)
display(df)



<ipython-input-64-c4c75cd7df5f>:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


name  count
0           Компьютерный мастер     79
1  Ученик компьютерного мастера     47

<ipython-input-64-c4c75cd7df5f>:32: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


name
0  None

**Вывод по ИП Поляков** - сильно понятнее не стало :-)